File: src/com/study/algorithm/algorithms/最小栈.java
package com.study.algorithm.algorithms;

import org.junit.Test;

import java.util.Comparator;
import java.util.PriorityQueue;

public class 最小栈 {
    @Test
    public void 最小栈() {
        MinStack minStack = new MinStack();
        minStack.push(-2);
        minStack.push(0);
        minStack.push(-3);
        minStack.getMin();
        minStack.pop();
        minStack.top();
        minStack.getMin();
    }

    public static class MinStack {

        MyNode head;

        MyNode min;
        PriorityQueue<MyNode> minStack;

        Comparator<MyNode> comparator =
                new Comparator<MyNode>() {
                    @Override
                    public int compare(MyNode o1, MyNode o2) {
                        if (o1.val > 0 && o2.val < 0) {
                            return 1;
                        } else if (o1.val < 0 && o2.val > 0) {
                            return -1;
                        } else {
                            return o1.val - o2.val;
                        }
                    }
                };

        /** initialize your data structure here. */
        public MinStack() {
            minStack = new PriorityQueue<>(comparator);
        }

        public void push(int x) {
            if (head == null) {
                head = new MyNode(x);
            } else {
                MyNode next = head;
                head = new MyNode(x);
                head.next = next;
            }
            minStack.offer(head);
        }

        public void pop() {
            MyNode next = head.next;
            minStack.remove(head);
            head.next = null;
            head = next;
        }

        public int top() {
            return head.val;
        }

        public int getMin() {
            return minStack.peek().val;
        }

        public static class MyNode {
            public int val;
            public MyNode next;

            public MyNode(int val) {
                this.val = val;
            }
        }
    }
}

File: src/com/study/algorithm/algorithmsKT/最小栈.kt
package com.study.algorithm.algorithmsKT

import java.util.*

class 最小栈 {

    //要使得操作的事件复杂度都是O(1)，那么可以使用一个辅助容器实时维护当前数组中最小的元素
    //每次加入新数据时，将此数据元素和辅助容器中当前数组下的最小元素进行比较并获取最小值后，再加入辅助容器中即可

    //哈希表存，当前数组大小下的最小元素(使用另一个栈亦可实时维护当前数组下的最小元素)
    val min = HashMap<Int, Int>()
    val stack = LinkedList<Int>()
    var size = 0

    fun push(`val`: Int) {
        stack.push(`val`)
        min[size + 1] = Math.min(min[size++] ?: `val`, `val`)
    }

    fun pop() {
        min.remove(size--)
    }

    fun top(): Int {
        return stack.peekFirst()
    }

    fun getMin(): Int {
        return min[size] ?: 0
    }

}